File: backend/app/models/fitting.py
"""
Pydantic Schemas for AI Precision Fit Pipeline
API Request / Response models matching Frontend & Database contracts (OpenAPI 3.0).
Supports bidirectional camelCase & snake_case serialization via Pydantic v2 alias generator.
"""

from typing import Any, Dict, Generic, List, Literal, Optional, TypeVar

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

T = TypeVar("T")


# ==============================================================================
# Base Model with CamelCase & SnakeCase Interoperability
# ==============================================================================
class CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        from_attributes=True
    )


# ==============================================================================
# Common Enums & Types
# ==============================================================================
GenderType = Literal["female", "male"]
FitPreferenceType = Literal["slim", "regular", "relaxed"]
BrandType = Literal["zara", "uniqlo", "hm", "pullandbear", "stradivarius"]
CategoryType = Literal["jacket", "coat", "blazer", "puffer", "dress", "shirt", "pants"]
SizeLabelType = Literal["XS", "S", "M", "L", "XL", "XXL"]
FabricStretchType = Literal["none", "low", "medium", "high"]


# ==============================================================================
# 0. Profile & Database Models (/api/v1/profiles)
# ==============================================================================
class UserProfile(CamelModel):
    id: str
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    name: str
    gender: GenderType
    fit_preference: FitPreferenceType
    height_cm: float = Field(..., ge=100.0, le=250.0)
    weight_kg: float = Field(..., ge=30.0, le=200.0)
    chest_cm: Optional[float] = Field(None, ge=50.0, le=160.0)
    waist_cm: Optional[float] = Field(None, ge=40.0, le=150.0)
    hips_cm: Optional[float] = Field(None, ge=50.0, le=160.0)
    shoulder_cm: Optional[float] = Field(None, ge=25.0, le=70.0)
    front_image_url: Optional[str] = None
    side_image_url: Optional[str] = None
    is_verified: bool = False
    is_default: bool = False
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CreateFitProfileRequest(CamelModel):
    name: str
    gender: GenderType
    fit_preference: FitPreferenceType
    height_cm: float = Field(..., ge=100.0, le=250.0)
    weight_kg: float = Field(..., ge=30.0, le=200.0)
    chest_cm: Optional[float] = None
    waist_cm: Optional[float] = None
    hips_cm: Optional[float] = None
    shoulder_cm: Optional[float] = None
    front_image_url: Optional[str] = None
    side_image_url: Optional[str] = None
    is_default: bool = False


class UpdateFitProfileRequest(CamelModel):
    name: Optional[str] = None
    gender: Optional[GenderType] = None
    fit_preference: Optional[FitPreferenceType] = None
    height_cm: Optional[float] = None
    weight_kg: Optional[float] = None
    chest_cm: Optional[float] = None
    waist_cm: Optional[float] = None
    hips_cm: Optional[float] = None
    shoulder_cm: Optional[float] = None
    front_image_url: Optional[str] = None
    side_image_url: Optional[str] = None
    is_verified: Optional[bool] = None
    is_default: Optional[bool] = None


class Garment(CamelModel):
    id: str
    sku: str
    name: str
    brand: BrandType
    category: CategoryType
    gender: Literal["female", "male", "unisex"]
    image_url: str
    fabric_stretch: FabricStretchType
    price: Optional[float] = None
    description: Optional[str] = None
    available_sizes: Optional[List[SizeLabelType]] = None


class GarmentSizeChart(CamelModel):
    id: str
    garment_id: str
    size: SizeLabelType
    chest_min: Optional[float] = None
    chest_max: Optional[float] = None
    waist_min: Optional[float] = None
    waist_max: Optional[float] = None
    hips_min: Optional[float] = None
    hips_max: Optional[float] = None
    shoulder_width: Optional[float] = None
    garment_length: Optional[float] = None


class GarmentDetail(Garment):
    size_charts: List[GarmentSizeChart]


# ==============================================================================
# 0.1 Pagination, Error & Response Standards
# ==============================================================================
class PaginationMetadata(CamelModel):
    page: int = Field(..., ge=1)
    page_size: int = Field(..., ge=1, le=100)
    total_items: int = Field(..., ge=0)
    total_pages: int = Field(..., ge=0)


class PaginatedResponse(CamelModel, Generic[T]):
    data: List[T]
    pagination: PaginationMetadata


class FitProfilesListResponse(CamelModel):
    data: List[UserProfile]
    total: int


class ErrorDetail(CamelModel):
    code: str
    message: str
    details: Optional[Any] = None


class ErrorResponse(CamelModel):
    error: ErrorDetail


# ==============================================================================
# 1. Quality Check API (/api/v1/quality-check)
# ==============================================================================
class QualityIssue(CamelModel):
    code: Literal["feet_cut_off", "head_cut_off", "blurry", "bad_lighting", "multi_person", "no_person", "bad_pose"]
    severity: Literal["error", "warning"]
    message: str = Field(..., description="Mô tả lỗi hiển thị cho người dùng (tiếng Việt)")
    box: Optional[List[float]] = Field(None, description="Bounding box vùng lỗi [ymin, xmin, ymax, xmax] normalized")


class QualityCheckRequest(CamelModel):
    image_base64: Optional[str] = None
    image_url: Optional[str] = None


class QualityCheckResponse(CamelModel):
    is_valid: bool
    confidence_score: float = Field(..., ge=0.0, le=1.0)
    issues: List[QualityIssue] = []
    blur_score: float = Field(..., description="Laplacian variance score")
    landmarks_detected: int = Field(..., description="Số lượng landmarks nhận diện được (tối đa 33)")


# ==============================================================================
# 2. Body Measurement API (/api/v1/measure)
# ==============================================================================
class BodyMeasurements(CamelModel):
    height_cm: float = Field(..., ge=100.0, le=250.0)
    weight_kg: float = Field(..., ge=30.0, le=200.0)
    shoulder_cm: Optional[float] = Field(None, ge=25.0, le=70.0)
    chest_cm: Optional[float] = Field(None, ge=50.0, le=160.0)
    waist_cm: Optional[float] = Field(None, ge=40.0, le=150.0)
    hips_cm: Optional[float] = Field(None, ge=50.0, le=160.0)


class MeasurementRequest(CamelModel):
    front_image_url: Optional[str] = None
    side_image_url: Optional[str] = None
    known_height_cm: Optional[float] = None
    gender: GenderType


class MeasurementResponse(CamelModel):
    measurements: BodyMeasurements
    confidence_percent: float = Field(..., ge=0.0, le=100.0)
    method: Literal["ai_vision", "anthropometric_hybrid"]


# ==============================================================================
# 3. Size Recommendation API (/api/v1/size-recommend)
# ==============================================================================
class PartFitEvaluation(CamelModel):
    part: Literal["shoulder", "chest", "waist", "hips", "length"]
    part_label: str = Field(..., description="Tên vị trí (Vai, Ngực, Eo, Hông, Dài áo)")
    status: Literal["perfect", "slightly_tight", "slightly_loose", "too_tight", "too_loose"]
    status_label: str = Field(..., description="Trạng thái trực quan: 'Vừa vặn', 'Hơi ôm', 'Hơi rộng'...")
    diff_cm: float = Field(..., description="Độ chênh lệch so với số đo người dùng (cm)")


class SizeComparisonItem(CamelModel):
    fit_score: float = Field(..., ge=0.0, le=1.0)
    badge_label: str = Field(..., description="'Vừa vặn chuẩn', 'Hơi ôm', 'Hơi rộng'...")
    description: str = Field(..., description="Giải thích ngắn gọn")


class SizeRecommendRequest(CamelModel):
    garment_id: str
    profile_id: Optional[str] = None
    fit_preference_override: Optional[FitPreferenceType] = None
    gender: Optional[GenderType] = None
    fit_preference: Optional[FitPreferenceType] = None
    measurements: Optional[BodyMeasurements] = None


class SizeRecommendResponse(CamelModel):
    recommended_size: SizeLabelType
    confidence_percent: int = Field(..., ge=0, le=100)
    fit_preference_label: str
    summary_explanation: str
    breakdown: List[PartFitEvaluation]
    size_comparisons: Dict[str, SizeComparisonItem]


# ==============================================================================
# 4. Virtual Try-On API (/api/v1/tryon)
# ==============================================================================
class CreateTryOnJobRequest(CamelModel):
    profile_id: str
    garment_id: str
    person_image_url: Optional[str] = None
    garment_image_url: Optional[str] = None


TryOnJobRequest = CreateTryOnJobRequest


class TryOnJob(CamelModel):
    job_id: str
    profile_id: Optional[str] = None
    garment_id: Optional[str] = None
    status: Literal["queued", "processing", "completed", "failed"]
    result_image_url: Optional[str] = None
    error_message: Optional[str] = None
    estimated_seconds: int = 15
    created_at: Optional[str] = None


TryOnJobResponse = TryOnJob

File: frontend/types/fitting.ts
/**
 * TypeScript Schemas for AI Precision Fit
 * Matching 1-1 with OpenAPI 3.0 Contract (docs/api/ai_precision_fit_api.yaml),
 * Backend Pydantic models (backend/app/models/fitting.py), & Supabase Tables.
 */

// =============================================================================
// Common Types & Enums
// =============================================================================
export type Gender = 'female' | 'male';
export type FitPreference = 'slim' | 'regular' | 'relaxed';
export type FitMethod = 'ai_photo' | 'manual';
export type SizeLabel = 'XS' | 'S' | 'M' | 'L' | 'XL' | 'XXL';
export type FabricStretch = 'none' | 'low' | 'medium' | 'high';
export type Brand = 'zara' | 'uniqlo' | 'hm' | 'pullandbear' | 'stradivarius';
export type Category = 'jacket' | 'coat' | 'blazer' | 'puffer' | 'dress' | 'shirt' | 'pants';

export type FlowStep =
  | 'welcome'
  | 'profile_setup'
  | 'method_select'
  | 'upload_guide'
  | 'upload_verify'
  | 'manual_input'
  | 'analyzing'
  | 'recommendation'
  | 'profile_list'
  | 'profile_detail';

// =============================================================================
// 1. Database & Core Model Schemas
// =============================================================================
export interface UserProfile {
  id: string;
  userId?: string | null;
  sessionId?: string;
  name: string;
  gender: Gender;
  fitPreference: FitPreference;
  heightCm: number;
  weightKg: number;
  chestCm?: number;
  waistCm?: number;
  hipsCm?: number;
  shoulderCm?: number;
  frontImageUrl?: string;
  sideImageUrl?: string;
  isVerified: boolean;
  isDefault: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export type FitProfile = UserProfile;

export interface CreateFitProfileRequest {
  name: string;
  gender: Gender;
  fitPreference: FitPreference;
  heightCm: number;
  weightKg: number;
  chestCm?: number;
  waistCm?: number;
  hipsCm?: number;
  shoulderCm?: number;
  frontImageUrl?: string;
  sideImageUrl?: string;
  isDefault?: boolean;
}

export interface UpdateFitProfileRequest {
  name?: string;
  gender?: Gender;
  fitPreference?: FitPreference;
  heightCm?: number;
  weightKg?: number;
  chestCm?: number;
  waistCm?: number;
  hipsCm?: number;
  shoulderCm?: number;
  frontImageUrl?: string;
  sideImageUrl?: string;
  isVerified?: boolean;
  isDefault?: boolean;
}

export interface Garment {
  id: string;
  sku: string;
  name: string;
  brand: Brand;
  category: Category;
  gender: Gender | 'unisex';
  imageUrl: string;
  fabricStretch: FabricStretch;
  price?: number;
  description?: string;
  availableSizes?: SizeLabel[];
}

export interface GarmentSizeChart {
  id: string;
  garmentId: string;
  size: SizeLabel;
  chestMin?: number;
  chestMax?: number;
  waistMin?: number;
  waistMax?: number;
  hipsMin?: number;
  hipsMax?: number;
  shoulderWidth?: number;
  garmentLength?: number;
}

export interface GarmentDetail extends Garment {
  sizeCharts: GarmentSizeChart[];
}

// =============================================================================
// 2. Pagination & API Response Standards
// =============================================================================
export interface PaginationMetadata {
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: PaginationMetadata;
}

export interface GarmentsQueryParams {
  page?: number;
  pageSize?: number;
  brand?: 'all' | Brand;
  category?: 'all' | Category;
  gender?: 'all' | Gender | 'unisex';
  search?: string;
}

export interface FitProfilesListResponse {
  data: UserProfile[];
  total: number;
}

export interface APIErrorDetail {
  loc?: (string | number)[];
  msg?: string;
  type?: string;
  [key: string]: unknown;
}

export interface APIErrorResponse {
  error: {
    code: string;
    message: string;
    details?: APIErrorDetail[] | unknown;
  };
}

// =============================================================================
// 3. AI Pipeline & Try-On Request & Response Schemas
// =============================================================================

// --- /api/v1/quality-check ---
export interface QualityIssue {
  code: 'feet_cut_off' | 'head_cut_off' | 'blurry' | 'bad_lighting' | 'multi_person' | 'no_person' | 'bad_pose';
  severity: 'error' | 'warning';
  message: string;
  box?: [number, number, number, number]; // [ymin, xmin, ymax, xmax]
}

export interface QualityCheckRequest {
  imageBase64?: string;
  imageUrl?: string;
}

export interface QualityCheckResponse {
  isValid: boolean;
  confidenceScore: number;
  issues: QualityIssue[];
  blurScore: number;
  landmarksDetected: number;
}

// --- /api/v1/measure ---
export interface BodyMeasurements {
  heightCm: number;
  weightKg: number;
  shoulderCm?: number;
  chestCm?: number;
  waistCm?: number;
  hipsCm?: number;
}

export interface MeasurementRequest {
  frontImageUrl?: string;
  sideImageUrl?: string;
  knownHeightCm?: number;
  gender: Gender;
}

export interface MeasurementResponse {
  measurements: BodyMeasurements;
  confidencePercent: number;
  method: 'ai_vision' | 'anthropometric_hybrid';
}

// --- /api/v1/size-recommend ---
export interface PartFitEvaluation {
  part: 'shoulder' | 'chest' | 'waist' | 'hips' | 'length';
  partLabel: string;
  status: 'perfect' | 'slightly_tight' | 'slightly_loose' | 'too_tight' | 'too_loose';
  statusLabel: string;
  diffCm: number;
}

export interface SizeComparisonItem {
  fitScore: number;
  badgeLabel: string;
  description: string;
}

export interface SizeRecommendRequest {
  garmentId: string;
  profileId?: string;
  fitPreferenceOverride?: FitPreference;
  gender?: Gender;
  fitPreference?: FitPreference;
  measurements?: BodyMeasurements;
}

export interface SizeRecommendResponse {
  recommendedSize: SizeLabel;
  confidencePercent: number;
  fitPreferenceLabel: string;
  summaryExplanation: string;
  breakdown: PartFitEvaluation[];
  sizeComparisons: Record<string, SizeComparisonItem>;
}

export type SizeRecommendationResult = SizeRecommendResponse;

// --- /api/v1/tryon ---
export interface CreateTryOnJobRequest {
  profileId: string;
  garmentId: string;
  personImageUrl?: string;
  garmentImageUrl?: string;
}

export type TryOnJobRequest = CreateTryOnJobRequest;

export interface TryOnJob {
  jobId: string;
  profileId?: string;
  garmentId?: string;
  status: 'queued' | 'processing' | 'completed' | 'failed';
  resultImageUrl?: string;
  errorMessage?: string;
  estimatedSeconds: number;
  createdAt?: string;
}

export type TryOnJobResponse = TryOnJob;
